tensorflow_datasets/scripts/cleanup/url_filename_recorder.py

```diff
@@ -0,0 +1,74 @@
+# coding=utf-8
+# Copyright 2020 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""Update checksums to contain filename."""
+
+import os
+import pathlib
+import requests
+from absl import app
+from typing import Dict, List
+from concurrent import futures
+
+import tensorflow_datasets as tfds
+from tensorflow_datasets.core.download import checksums
+
+
+TFDS_PATH = tfds.core.utils.tfds_write_path()
+
+
+def _collect_path_to_update() -> List[tfds.core.utils.ReadWritePath]:
+  """Collect checksums paths to update."""
+  url_checksums_paths = list(checksums._checksum_paths().values())
+
+  builder_checksums_path = []
+  for name in tfds.list_builders():
+    if tfds.builder_cls(name).url_infos != None:
+      builder_checksums_path.append(tfds.builder_cls(name)._checksums_path)
+  return url_checksums_paths + builder_checksums_path
+
+def _request_filename(url) -> (str, str):
+  """Get filename of dataset at `url`."""
+  filename  = ''
+  try:
+    response = requests.get(url, timeout=10)
+    filename = tfds.core.download.downloader._get_filename(response)
+    print(f'Success for {url}')
+  except requests.exceptions.HTTPError as http_err:
+    print(f'HTTP Error {http_err} for {url}.')
+  except Exception as error:
+    print(f'Error {error} for {url}.')
+  return (url, filename)
+
+def _update_url_infos(url_infos) -> Dict[str, checksums.UrlInfo]:
+  """Get and update dataset filname in UrlInfo."""
+  with futures.ThreadPoolExecutor(max_workers=100) as executor:
```

Comment:
Note: 
Here you're recreating a new thread pool for each datasets. So each dataset is processed individually. It would be much faster to share the same threadpool across all datasets, like:

```py
with futures.ThreadPoolExecutor(max_workers=100) as executor:
    path_to_new_url_infos = {}
    # Add request to the executor queue
    for url_path in _collect_path_to_update():
      url_infos = checksums.load_url_infos(url_path)
      path_to_new_url_infos[url_path] = executor.map(_request_filename, url_infos)
    # Process results
    for path, url_info_futures in path_to_new_url_infos.items():
      ...
```

Reply:
I thought to do so first, but then while implementing I realised that for separately processing the results of `Future` instances, I need to either reload the `url_infos` or cache the entire thing. So, instead I concurrently loaded and updated urls within each datasets.
No worries, I will do this in the next commit. 

Thanks.

Reply:
Oh yes, `url_infos` should be cached indeed. This is not an issue. `ckecksums.get_all_url_infos()` already cache all url_infos for all datasets

Reply:
Then this could be: `path_to_url_infos = _collect_path_to_url_infos() -> Dict[path, Dict[url, UrlInfo]]`.

There is still the issue that multiple datasets might use the same url which will currently downloaded twice. Let me know if you're interested by this but that might be over-optimisation.

Reply:
Thats a great idea!

Thanks.